src/handlers.rs

```diff
@@ -111,11 +114,106 @@ pub fn handle_request(log: &Logger, request: &protocol::Request) -> Result<Vec<u
             };
             serialize_initgroups(groups)
         }
-        RequestType::GETHOSTBYADDR
-        | RequestType::GETHOSTBYADDRv6
-        | RequestType::GETHOSTBYNAME
-        | RequestType::GETHOSTBYNAMEv6
-        | RequestType::SHUTDOWN
+
+        // GETHOSTBYADDR and GETHOSTBYADDRv6 implement reverse lookup
+        // The key contains the address to look for.
+        RequestType::GETHOSTBYADDR => {
+            let key = request.key;
+
+            if key.len() != 4 {
+                bail!("Invalid key len: {}, expected 4", key.len());
+            }
+            let address_bytes: [u8; 4] = key.try_into()?;
+            let address = IpAddr::from(address_bytes);
+
+            let sock = SocketAddr::new(address, 0);
+            let host = match getnameinfo(&sock, NI_NUMERICSERV) {
+                Ok((hostname, _service)) => Ok(Some(Host {
+                    addresses: vec![address],
+                    hostname,
+                })),
+                Err(e) => match e.kind() {
+                    dns_lookup::LookupErrorKind::NoName => Ok(None),
+                    _ => bail!("error during lookup: {:?}", e),
+                },
+            };
+            Ok(serialize_host(host))
+        }
+        RequestType::GETHOSTBYADDRv6 => {
+            let key = request.key;
+
+            if key.len() != 16 {
+                bail!("Invalid key len: {}, expected 16", key.len());
+            }
+            let address_bytes: [u8; 16] = key.try_into()?;
+            let address = IpAddr::from(address_bytes);
+
+            let sock = SocketAddr::new(address, 0);
+            let host = match getnameinfo(&sock, NI_NUMERICSERV) {
+                Ok((hostname, _service)) => Ok(Some(Host {
+                    addresses: vec![address],
+                    hostname,
+                })),
+                Err(e) => match e.kind() {
+                    dns_lookup::LookupErrorKind::NoName => Ok(None),
+                    _ => bail!("error during lookup: {:?}", e),
+                },
+            };
+            Ok(serialize_host(host))
+        }
+
+        RequestType::GETHOSTBYNAME => {
+            let hostname = CStr::from_bytes_with_nul(request.key)?.to_str()?;
+            let hints = AddrInfoHints {
+                socktype: SOCK_STREAM,
+                ..AddrInfoHints::default()
+            };
+
+            let host = match getaddrinfo(Some(hostname), None, Some(hints)) {
+                Ok(addrs) => {
+                    let addresses: std::io::Result<Vec<_>> =
+                        addrs.map(|r| r.map(|a| a.sockaddr.ip())).collect();
```

Comment:
Here (and in the v6 case further down) I forgot to filter, and in case both address families are returned, `serialize_hosts` bails out silently with an error (because we have no logger there).

I'll send a fix for this in a bit.

Reply:
Done!